File: backend/app/crypto_odds.py
"""Crypto up/down odds — log-normal model with platform margin (shared by API + SSE)."""
from __future__ import annotations

import math

CRYPTO_MARGIN = 0.70
CRYPTO_MIN_ODDS = 1.1
# Annualized σ for log-normal scaling; tuned above spot vol so short-window
# ticks (often ≪0.1%) do not swing Higher/Lower quotes on noise alone.
CRYPTO_BTC_VOL = 1.60
SECS_PER_YEAR = 365.25 * 24 * 3600

# Must match frontend `cryptoWindowClock` / `btc_price._PERIOD_MS` (+ 24h).
WINDOW_SEC: dict[str, float] = {"5m": 300.0, "30m": 1800.0, "24h": 86400.0}
# Fraction of window left → scale on σ; higher = odds lock in harder near the buzzer.
TIME_URGENCY_EXP = 0.62
MIN_TIME_FRAC = 0.02


def _norm_cdf(x: float) -> float:
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2)))


def window_seconds(window: str | None) -> float | None:
    if not window:
        return None
    return WINDOW_SEC.get(window)


def time_urgency_scale(remaining_sec: float, window_sec: float | None) -> float:
    """<1 when the clock is low — less σ left → current lead matters more."""
    if window_sec is None or window_sec <= 0:
        return 1.0
    frac = max(MIN_TIME_FRAC, min(1.0, remaining_sec / window_sec))
    return frac**TIME_URGENCY_EXP


def time_fraction(remaining_sec: float, window_sec: float | None) -> float | None:
    if window_sec is None or window_sec <= 0:
        return None
    return max(MIN_TIME_FRAC, min(1.0, remaining_sec / window_sec))


def calc_crypto_odds_with_vol(
    annual_vol: float,
    current: float,
    openp: float,
    remaining_sec: float,
    *,
    window_sec: float | None = None,
) -> dict[str, float]:
    if current <= 0 or openp <= 0 or remaining_sec <= 0:
        return {"up": 1.40, "down": 1.40}
    vol = max(0.01, float(annual_vol))
    T = remaining_sec / SECS_PER_YEAR
    t_scale = time_urgency_scale(remaining_sec, window_sec)
    sig_sqrt_t = max(1e-9, vol * math.sqrt(T) * t_scale)
    d = math.log(current / openp) / sig_sqrt_t
    p_up = max(0.01, min(0.99, _norm_cdf(d)))
    p_down = 1 - p_up
    return {
        "up": round(min(15, max(1.01, CRYPTO_MARGIN / p_up)), 2),
        "down": round(min(15, max(1.01, CRYPTO_MARGIN / p_down)), 2),
    }


def calc_crypto_odds(
    current: float,
    openp: float,
    remaining_sec: float,
    *,
    window: str | None = None,
) -> dict[str, float]:
    return calc_crypto_odds_with_vol(
        CRYPTO_BTC_VOL,
        current,
        openp,
        remaining_sec,
        window_sec=window_seconds(window),
    )

File: frontend/src/lib/cryptoOdds.js
/**
 * Crypto odds — must stay aligned with `backend/app/crypto_odds.py` + fair-odds service.
 * Server is source of truth; client recalculates every tick using `fairOdds.effectiveVol`
 * and live `remainingSec` (time left is a major driver of quotes).
 */

export const CRYPTO_MARGIN = 0.7
export const CRYPTO_BTC_VOL = 1.6
/** Minimum multiplier to accept a crypto bet (must match backend). */
export const CRYPTO_MIN_ODDS = 1.1
const SECS_PER_YEAR = 365.25 * 24 * 3600

/** @type {Record<string, number>} */
export const CRYPTO_WINDOW_SEC = { '5m': 300, '30m': 1800, '24h': 86400 }

const TIME_URGENCY_EXP = 0.62
const MIN_TIME_FRAC = 0.02

function erf(x) {
  const sign = x >= 0 ? 1 : -1
  const ax = Math.abs(x)
  const t = 1 / (1 + 0.3275911 * ax)
  const y =
    1 -
    (((((1.061405429 * t - 1.453152027) * t + 1.421413741) * t - 0.284496736) * t +
      0.254829592) *
      t *
      Math.exp(-ax * ax))
  return sign * y
}

function normCdf(x) {
  return 0.5 * (1 + erf(x / Math.SQRT2))
}

export function cryptoWindowSeconds(windowKey) {
  if (!windowKey) return null
  return CRYPTO_WINDOW_SEC[windowKey] ?? null
}

export function timeUrgencyScale(remainingSec, windowSec) {
  if (!windowSec || windowSec <= 0) return 1
  const frac = Math.max(MIN_TIME_FRAC, Math.min(1, remainingSec / windowSec))
  return frac ** TIME_URGENCY_EXP
}

/**
 * @param {number} annualVol
 * @param {number} current
 * @param {number} openp
 * @param {number} remainingSec
 * @param {number | null | undefined} windowSec
 * @returns {{ up: number, down: number }}
 */
export function calcCryptoOddsWithVol(annualVol, current, openp, remainingSec, windowSec) {
  if (current <= 0 || openp <= 0 || remainingSec <= 0) {
    return { up: 1.40, down: 1.40 }
  }
  const vol = Math.max(0.01, Number(annualVol))
  const T = remainingSec / SECS_PER_YEAR
  const tScale = timeUrgencyScale(remainingSec, windowSec)
  const sigSqrtT = Math.max(1e-9, vol * Math.sqrt(T) * tScale)
  const d = Math.log(current / openp) / sigSqrtT
  const pUp = Math.max(0.01, Math.min(0.99, normCdf(d)))
  const pDown = 1 - pUp
  const quote = (p) =>
    Math.round(Math.min(15, Math.max(1.01, CRYPTO_MARGIN / p)) * 100) / 100
  return { up: quote(pUp), down: quote(pDown) }
}

/**
 * @param {string | null | undefined} windowKey
 * @param {Record<string, unknown> | null | undefined} serverOdds
 * @param {Record<string, unknown> | null | undefined} fairOdds
 * @param {number | null | undefined} current
 * @param {number | null | undefined} openp
 * @param {number} remainingSec
 */
export function pickCryptoOdds(windowKey, serverOdds, fairOdds, current, openp, remainingSec) {
  if (current == null || openp == null || remainingSec <= 0) {
    return { up: null, down: null, source: 'none' }
  }
  const windowSec =
    cryptoWindowSeconds(windowKey) ??
    (typeof fairOdds?.windowSec === 'number' ? fairOdds.windowSec : null)

  const eff = fairOdds?.effectiveVol
  if (typeof eff === 'number' && Number.isFinite(eff) && eff > 0) {
    const o = calcCryptoOddsWithVol(eff, current, openp, remainingSec, windowSec)
    return { ...o, source: 'dynamic' }
  }

  const up = serverOdds?.up
  const down = serverOdds?.down
  if (
    typeof up === 'number' &&
    typeof down === 'number' &&
    Number.isFinite(up) &&
    Number.isFinite(down)
  ) {
    return { up, down, source: 'server' }
  }

  const local = calcCryptoOdds(current, openp, remainingSec, windowKey)
  return { ...local, source: 'local' }
}

/** @param {string | null | undefined} windowKey */
export function calcCryptoOdds(current, openp, remainingSec, windowKey) {
  return calcCryptoOddsWithVol(
    CRYPTO_BTC_VOL,
    current,
    openp,
    remainingSec,
    cryptoWindowSeconds(windowKey),
  )
}

export function isCryptoOddsBettable(n) {
  return n != null && Number.isFinite(n) && n >= CRYPTO_MIN_ODDS
}

export function fmtCryptoOddsMult(n) {
  if (!isCryptoOddsBettable(n)) return '—'
  return (
    n.toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 }) + 'x'
  )
}

/** Parse `crypto-k-btc-5m` → `5m`. */
export function cryptoWindowFromBetId(betId) {
  const id = String(betId).replace(/^crypto-/, '')
  const m = id.match(/^k-btc-(.+)$/)
  return m ? m[1] : null
}
